routers/repo/repo.go

```diff
@@ -458,80 +455,79 @@ func RedirectDownload(ctx *context.Context) {
 	ctx.Error(404)
 }
 
-// Download download an archive of a repository
-func Download(ctx *context.Context) {
-	var (
-		uri         = ctx.Params("*")
-		refName     string
-		ext         string
-		archivePath string
-		archiveType git.ArchiveType
-	)
+// DownloadStatus checks the status of a download, because archiving may take a
+// while.  It does so by creating an archive request from the archiver service,
+// then just examining the completion status.
+func DownloadStatus(ctx *context.Context) {
+	uri := ctx.Params("*")
+	aReq := archiver_service.DeriveRequestFrom(ctx, uri)
 
-	switch {
-	case strings.HasSuffix(uri, ".zip"):
-		ext = ".zip"
-		archivePath = path.Join(ctx.Repo.GitRepo.Path, "archives/zip")
-		archiveType = git.ZIP
-	case strings.HasSuffix(uri, ".tar.gz"):
-		ext = ".tar.gz"
-		archivePath = path.Join(ctx.Repo.GitRepo.Path, "archives/targz")
-		archiveType = git.TARGZ
-	default:
-		log.Trace("Unknown format: %s", uri)
+	if aReq == nil {
 		ctx.Error(404)
 		return
 	}
-	refName = strings.TrimSuffix(uri, ext)
 
-	if !com.IsDir(archivePath) {
-		if err := os.MkdirAll(archivePath, os.ModePerm); err != nil {
-			ctx.ServerError("Download -> os.MkdirAll(archivePath)", err)
-			return
+	complete := aReq.IsComplete()
+	ctx.JSON(200, map[string]interface{}{
+		"archiving": !complete,
+		"complete":  complete,
+	})
+}
+
+// Download an archive of a repository
+func Download(ctx *context.Context) {
+	uri := ctx.Params("*")
+	aReq := archiver_service.DeriveRequestFrom(ctx, uri)
+
+	downloadName := ctx.Repo.Repository.Name + "-" + aReq.GetArchiveName()
+	if !aReq.IsComplete() {
+		aReq = archiver_service.ArchiveRepository(aReq)
+		archiver_service.LockQueue()
+		for !aReq.IsComplete() {
+			archiver_service.WaitForCompletion()
 		}
+		archiver_service.UnlockQueue()
 	}
```

Comment:
I think this block can be moved to the archiver service, so the lock semantics remain opaque.